lightning/src/routing/router.rs

```diff
@@ -1785,10 +1782,82 @@ fn add_random_cltv_offset(route: &mut Route, payment_params: &PaymentParameters,
 	}
 }
 
+/// Construct a route from us (payer) to the target node (payee) via the given hops (which should
+/// exclude the payer, but include the payee). This may be useful, e.g., for probing the chosen path.
+///
+/// Re-uses logic from `find_route`, so the restrictions described there also apply here.
+pub fn build_route_from_hops<L: Deref>(
+	our_node_pubkey: &PublicKey, hops: &[PublicKey], route_params: &RouteParameters, network: &NetworkGraph,
+	logger: L, random_seed_bytes: &[u8; 32]
+) -> Result<Route, LightningError>
+where L::Target: Logger {
+	let network_graph = network.read_only();
+	let mut route = build_route_from_hops_internal(
+		our_node_pubkey, hops, &route_params.payment_params, &network_graph,
+		route_params.final_value_msat, route_params.final_cltv_expiry_delta, logger, random_seed_bytes)?;
+	add_random_cltv_offset(&mut route, &route_params.payment_params, &network_graph, random_seed_bytes);
+	Ok(route)
+}
+
+fn build_route_from_hops_internal<L: Deref>(
+	our_node_pubkey: &PublicKey, hops: &[PublicKey], payment_params: &PaymentParameters,
+	network_graph: &ReadOnlyNetworkGraph, final_value_msat: u64, final_cltv_expiry_delta: u32,
+	logger: L, random_seed_bytes: &[u8; 32]
+) -> Result<Route, LightningError> where L::Target: Logger {
+
+	struct HopScorer<'a> {
+		our_node_id: &'a NodeId,
+		hop_ids: &'a [Option<NodeId>],
+	}
+
+	impl Score for HopScorer<'_> {
+		fn channel_penalty_msat(&self, _short_channel_id: u64, source: &NodeId, target: &NodeId,
+			_usage: ChannelUsage) -> u64
+		{
+			let mut cur_id = self.our_node_id;
+			for i in 0..self.hop_ids.len() {
+				if let Some(ref next_id) = self.hop_ids[i] {
```

Comment:
Should we `break` if this is `None`?

Reply:
Ah, sure! (https://github.com/lightningdevkit/rust-lightning/pull/1491/commits/2010670ac153ee1aeb4ccc40ee64a8b86ed54397).